packages/hardhat/contracts/strategy/CrabMigration.sol

```diff
@@ -0,0 +1,162 @@
+         crabV1 = CrabStrategy(_crabV1);
+         crabV2 = CrabStrategyV2 (_crabV2);
+         euler = IEulerExec(_eulerExec);
+         owner = msg.sender;
+         EULER_MAINNET = _eulerMainnet;
+         weth = WETH9(_weth);
+         dToken = _dToken; 
+         wPowerPerp = crabV2.wPowerPerp();
+     }
+
+
+     /** 
+     * @notice allows users to deposit their crab v1 shares in the pool for migration
+     */
+     function depositV1Shares(uint256 amount) external beforeMigration { 
+         sharesDeposited[msg.sender] += amount;
+         totalCrabV1SharesMigrated += amount;
+         crabV1.transferFrom(msg.sender, address(this), amount);
+
+     }
+
+     /**
+      * @notice the owner batch migrates all the crab v1 shares in this contract to crab v2 and initializes 
+      * the v2 contract at the same collateral ratio as the v1 contract. 
+      */
+     function batchMigrate() external onlyOwner beforeMigration { 
+        // 1. update isMigrated
+        isMigrated = true;
+
+         // 2. flash floan eth from euler eq to amt 
+        bytes memory data;
+        euler.deferLiquidityCheck(address(this), data);
+
+        // 3. record totalV2Shares
+        totalCrabV2SharesReceived = crabV2.balanceOf(address(this));
+
+     }
+
+     function onDeferredLiquidityCheck(bytes memory encodedData) external {
+        require(msg.sender == EULER_MAINNET, "M4"); 
+
+        // 1. Borrow weth
+        uint256 crabV1Balance = crabV1.balanceOf(address(this));
+        uint256 crabV1Supply = crabV1.totalSupply();
+        (address _, uint256 id, uint256 totalCollateral, uint256 totalShort) = crabV1.getVaultDetails();
+        uint256 amountEthToBorrow = totalCollateral.wmul(crabV1Balance.wdiv(crabV1Supply));
+
+        IEulerDToken(dToken).borrow(0, amountEthToBorrow);
+        weth.withdraw(amountEthToBorrow);
+
+        // 2. mint osqth in crab v2 
+        uint256 wSqueethToMint = crabV1.getWsqueethFromCrabAmount(crabV1Balance);
+        crabV2.initialize{value: amountEthToBorrow}(wSqueethToMint);
+
+        // 3. call withdraw from crab v1
+        IERC20(wPowerPerp).approve(address(crabV1), type(uint).max);
+        crabV1.approve(address(crabV1), crabV1Balance);
+        crabV1.withdraw(crabV1Balance);
+
+        // 4. Repay the weth:
+        weth.deposit{value: amountEthToBorrow}();
+        weth.approve(EULER_MAINNET, type(uint).max);
+        IEulerDToken(dToken).repay(0, amountEthToBorrow);
+    }
+
+     /**
+      * @notice allows users to claim their amount of crab v2 shares
+      */
+     function claimV2Shares() external afterMigration { 
+         uint256 amountV1Deposited = sharesDeposited[msg.sender];
+         sharesDeposited[msg.sender] = 0;
+         uint256 crabV2TotalShares = crabV2.balanceOf(address(this));
+         uint256 amountV2ToTransfer = amountV1Deposited * totalCrabV2SharesReceived / totalCrabV1SharesMigrated;
```

Comment:
How does this go with precision? Could end up taking 1 more or less from rounding?